---
title: "Exploratory Analysis"
author: "Shubham Gupta"
date: "July 31, 2016"
output: html_document
---

Clean the global environment
```{r Clean_Environment}
rm(list = ls())
```

Assign colors to global palette
```{r GlobalPalette}
tropical=  c('darkorange', 'dodgerblue', 'hotpink', 'limegreen', 'yellow')
palette(tropical)
par(pch = 21)
```

Look at this link to know more about pch http://www.statmethods.net/advgraphs/parameters.html

Uncomment to install packages    
## ----install_packages, eval=FALSE--------------------------------     
#  install.packages(c("devtools","gplots"))      
#  source("http://www.bioconductor.org/biocLite.R")     
#  biocLite(c("Biobase","org.Hs.eg.db","AnnotationDbi"))    
#  biocLite("alyssafrazee/RSkittleBrewer")   
#  install.package("dendextend")

```{r load_hidden, echo=FALSE, results="hide"}
suppressPackageStartupMessages({
  library(devtools)
  library(Biobase)
  library(RSkittleBrewer)
  library(gplots)
  library(dplyr)
  library(org.Hs.eg.db)
  library(AnnotationDbi)
  library(dendextend)
})
```

### Load libraries
```{r loadLibs, cache = TRUE}
library(gplots)
library(devtools)
library(Biobase)
library(RSkittleBrewer)
library(org.Hs.eg.db)
library(AnnotationDbi)
library(dendextend)
```

## Load the data
```{r loadData, cache = TRUE}
con = url("http://bowtie-bio.sourceforge.net/recount/ExpressionSets/bodymap_eset.RData")
load(file=con)
close(con)
bm = bodymap.eset
pdata=pData(bm)
edata=exprs(bm)
fdata = fData(bm)
ls()
```

### Understand data
Use str, head, class, summary, dim command to understand properties and nature of data before executing following commands
```{r ObserveData, eval=FALSE}
## ----tables--------------------------------------------------------------
table(pdata$gender)
table(pdata$gender,pdata$race)

## ----summary: You can see that data is skewed by looking at min, max and mean ------
summary(edata)

## ----missing-------------------------------------------------------------
# Use option useNA to include NA's in table
table(pdata$age,useNA="ifany")

# is.na checks for NA values
table(is.na(pdata$age)) # Gives 3 NA values

# Check for other common missing names
sum(pdata$age==" ")
sum(pdata$age==" ", na.rm = TRUE) # So no missing values

# Check genomic data for NAs
sum(is.na(edata))

# Make the distribution of NA's by genes
gene_na = rowSums(is.na(edata))
class(gene_na)
dim(gene_na)
str(gene_na)
head(gene_na)
summary(gene_na)
table(gene_na)

# Make the distribution of NA's by samples
sample_na = rowSums(is.na(edata))
table(sample_na)
```

Check to make sure dimensions match
```{r dimension}
dim(fdata)
dim(pdata)
dim(edata)
```

### Graphical view of data
First we make box plot
```{r figures, figure.align = 'center', cache = TRUE}
boxplot(edata[,1])
boxplot(log2(edata[,1]+1)) # +1 because some values are zero
boxplot(log2(edata + 1), col = 2, range = 0)
par(mfrow = c(1,2)) # To have 2 plots in one row
hist(log2(edata[,1] + 1), col = 1)
hist(log2(edata[,2] + 1), col = 2)
par(mfrow = c(1,1))
plot(density(log2(edata[,1]+1)), col =1)
# lines allows to have overlaying graphs
lines(density(log2(edata[,1]+1)), col =1)
lines(density(log2(edata[,2]+1)), col =2)
```

With lines plot, we can see if there is any artifact when lines don't overlap and we see a strange bump or something telling that something went awry.

Other types of figures:    
Quantile- quantile plots (QQ plot)    
MA plot    
```{r, figures1, figure.align = 'center', cache = TRUE}
qqplot(log2(edata[,1]+1), log2(edata[,2]+1),col=3)
abline(0,1)

mm = log2(edata[,1]+1) - log2(edata[,2]+1)
aa = log2(edata[,1]+1) + log2(edata[,2]+1)
plot(aa,mm,col=2)
```


Since there are so many low values in edata we need to filter them. Therefore, we need to first transform dataset from matrix to dataframe so that we can apply dplyr filters.    
```{r filerData, warning= FALSE, cache = TRUE}
class(edata)
edata = as.data.frame(edata)
class(edata)
filt_edata = filter(edata, rowMeans(edata)>1)
dim(filt_edata)
boxplot(as.matrix(log2(filt_edata+1)), col = 4)
```

### Check for consistency
Check for consistency using some external data     
org.Hs.eg.db = Genome wide annotation for Human    
keys must be a character so we need to do casting     
```{r ChrInfo, cache = TRUE}
class(fdata[,1])
aeid = as.character(fdata[,1])
# Extract chromosome information from ENSEMBL ids
chr = AnnotationDbi::select(org.Hs.eg.db,keys=aeid,keytype="ENSEMBL",columns="CHR")
class(chr)
dim(chr)
dim(fdata)
colnames(chr)
head(chr)
# Take non-duplicated chromsomes
chr = chr[!duplicated(chr[,1]),]
dim(chr)
all(chr[,1] == rownames(fdata))
all(chr[,1] == rownames(edata))

```


## Data transformation
Log is good transformation variable for highly skewed data    
We take log2 because when we compare them, we subtract and now if we see value increased by 1 means value is doubled.
```{r DataTransform, cache=TRUE}
head(edata,n = 5)
hist(edata[,1], col =6)
hist(log2(edata[,1]+1), col =6)
hist(log2(edata[,1]+1), col =6, breaks = 1000)
hist(log(edata[,1]+1), col =6, breaks = 1000)
#Zoom in to ignore zeros
hist(log2(edata[,1]+1), col =2, breaks = 100, xlim = c(1,15), ylim = c(0,300))
hist(rowSums(edata==0),col =3)
```

rowSums sums along the row, so it combines TRUE value for all 19 columns. And from the graph we can see that there are many rows(genes) which has value(expression value) = 0 in all columns(patients)

Now we remove the zero values.Filter low expressed genes.
```{r FiltGene, cache= TRUE}
low_genes  = rowMeans(as.matrix(edata))<5;
table(low_genes)
filt_edata = filter(edata, !low_genes)
hist(log2(filt_edata[,1]+1), col = 3)
low_genes2  = rowMedians(as.matrix(edata))<5;
table(low_genes2)
table(low_genes2, low_genes)
filt_edata = filter(edata, !low_genes2)
hist(log2(filt_edata[,1]+1), col = 4)
```

## Clustering analysis
First install "dendextend" package and load library



### Session Info
```{r sessionInfo, cache = TRUE}
devtools::session_info()
```

### Last compilation

Last compiled at `r Sys.Date()`.
